### PR TITLE
Alan create new user duplicate name fix

### PR DESCRIPTION
**Pull request by LeeSingChanYau**
This is a PR for Other Links → User Management → Create New User ([PR 476](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/476) needs finishing, a “duplicate name created” popup happens after it is created, which is good. Need the rest of the below still though)
Make "user with duplicate name" popup require a person to click to confirm that they want to create a person with a duplicate name 
Make "user with duplicate name" so it is not case sensitive. Currently a user without any caps is not considered a duplicate of the name that includes capitalization of first names. 
The above PR also seems to have broken the auto-refresh for adding a new person… I should see that new person as soon as they are added and now I don’t anymore

How to test:
- With a dev account go to Other Links → User Management → Create New User.
- Create a user with with a name that has been used before (first name + last name).
- Create it with an email that hasn't been used before and also with one that has been used before.
- A new popup should appear asking to confirm if you want to create the user with a duplicate name.
- Check if it behaves the right way.
- Check if the newly created user shows up.

### CODE REVIEW

**Review by rauleffting — COMMENTED** [review]
Hi @LeeSingChanYau, I tested your PR and worked quite well:

Screen Shot 2023-01-17 at 13 03 50

Screen Shot 2023-01-17 at 13 04 21

However, I have one question: why did you styled inside the html tags instead using a css file?

Thank you very much!

File: src/components/UserManagement/DuplicateNamePopup.jsx
Comment (rauleffting) [inline src/components/UserManagement/DuplicateNamePopup.jsx:19]: Here, for example.

Comment (LeeSingChanYau) [reply src/components/UserManagement/DuplicateNamePopup.jsx:19]: Hey Raul, this is because I'm following the same convention as other popups that are already created. Note that we are using the reactstrap package for buttons so it is not the same as a regular button. I tried using the traditional way by importing the css, but it does not look the same.

Comment (LeeSingChanYau) [reply src/components/UserManagement/DuplicateNamePopup.jsx]: I just did a new commit with small changes to the style of the button and doesn't use any css modifications. Please review.

Comment (LeeSingChanYau) [reply src/components/UserManagement/DuplicateNamePopup.jsx]: Screenshot 2023-01-17 at 10 07 57 AM

**Review by rauleffting — DISMISSED** [review]
Perfect, Allan! I tested again. The code is cleaner now.

Thank you very much!

I'm going to approve it.

**Comment by RafaelCastilhoCastro** [comment]
Hi @LeeSingChanYau,

I tested the PR but I got different results from what you and Raul described. I'm not getting any confirmation pop up. Instead I'm getting screen notifications alerting that duplicate users are being created.
1. When creating a new user with an existing email the alert is red saying the email is already in use (Check attached img 1). The modal stays open awaiting for the corrections.
2. When creating a new user with an existing name but with a new email the alert is yellow saying a duplicate user was created (Check image 2). The modal is closed and the duplicated user shows on the list.



For this test I only did checkout to this PR. I didn't checkout to any specific PR on Backend. Please let me know if I should do something different.

**Comment by RafaelCastilhoCastro** [comment]
Updating my comment:
I think this PR is related to Backend PR #272, according to Alan's initial message. I did checkout to PR #272 on Backend and tested it again. It worked as intended.
1. When creating a new user using an existing email, the same red alert pops and the modal stays open awaiting to be corrected.
2. When creating a new user with the same name but different email, a confirmation modal pops up.

**Review by RJBornToWin — CHANGES_REQUESTED** [review]
reviewing.

File: src/components/UserManagement/DuplicateNamePopup.jsx
Comment (PrabhjotSembhi) [inline src/components/UserManagement/DuplicateNamePopup.jsx:28]: I think we can use "Change name" rather than "Cancel".  It will be more clear.

You can also set focus to name fields in the form. (optional)

**Review by PrabhjotSembhi — CHANGES_REQUESTED** [review]
I have checked by registering users with the same first and last name. with the same and different. it's working the way we want.

it's asking for a different email if it's already in use. and the duplicate name alert popup is working correctly.



But We are accepting the same Phone number for different accounts.  
I suggest you should address this also, the same way you did with email. 
Alert the user if a Phone number is already in use. 
image
image


You can check the video and screenshots for more clarity.


https://user-images.githubusercontent.com/82662768/214269047-61059bb8-1709-49b3-8ed3-281167fa3ce0.mp4

File: src/components/UserManagement/DuplicateNamePopup.jsx
Comment (one-community) [reply src/components/UserManagement/DuplicateNamePopup.jsx:28]: Good catch Prabhjot! We're allowing the same phone number because fake data is sometimes used for all the testing, and the odds of the same number actually being entered by accident for two people is so low. Some orgs may not care about the number too, and will add ALL fake info... so we don't want that to create problems for orgs like us that use real info.

**Comment by LeeSingChanYau** [comment]
Hey @PrabhjotSembhi, yes that would be a good feature. I'll confirm first if we want to check duplicate phone numbers.

File: src/components/UserManagement/DuplicateNamePopup.jsx
Comment (LeeSingChanYau) [reply src/components/UserManagement/DuplicateNamePopup.jsx]: Hey @PrabhjotSembhi, as @one-community said we won't check for duplicate phone numbers. I did update the text of the button from "Cancel" to "Change name".

Comment (one-community) [reply src/components/UserManagement/DuplicateNamePopup.jsx]: Thank you both

**Comment by AnishPandita** [comment]
Hey @LeeSingChanYau , Tested your PR with the necessary backend branch and everything seems to work fine. Overall code looks neat too.

https://user-images.githubusercontent.com/111321945/215164597-945ecb78-87e3-4b8c-bbfb-dc54747df26d.mov

**Comment by Gukal7** [comment]
Hey @LeeSingChanYau, I have tested the code and everything looks perfect.
Screenshot 2023-01-27 at 12 23 02 PM
Screenshot 2023-01-27 at 12 22 34 PM
Screenshot 2023-01-27 at 12 22 19 PM

**Review by PrabhjotSembhi — COMMENTED** [review]
Now its good

**Review by PrabhjotSembhi — DISMISSED** [review]
As we dont want to check for the same phone number. 
Email check is working well. 
I am approving this.

**Review by RJBornToWin — CHANGES_REQUESTED** [review]
I have tested the code. The basic functionality fails. Please refer the video. Thanks.


https://user-images.githubusercontent.com/16258136/215933216-b3c77cd1-680b-4df4-9c69-1b9849a8b07a.mov

**Comment by RJBornToWin** [comment]
@PrabhjotSembhi @Gukal7 @AnishPandita : Could you please check my workflow and see if it works fine for you? Thanks.
FYI: @rauleffting

**Comment by LeeSingChanYau** [comment]
@RJBornToWin Changed if(duplicates) to if(duplicates.length > 0). This should fix the problem. Please review.

**Comment by RafaelCastilhoCastro** [comment]
Hi @LeeSingChanYau,
I tested your PR again and observed an issue I hadn't experienced before.

1. First I tryed to replicate the bug reported by @RJBornToWin. But in my case the feature worked as intended as you can see in the attached video.

https://user-images.githubusercontent.com/65736872/216771709-00579bbf-e898-4586-a125-59ed82ba5fa7.mp4



2. Now I'm experiencing the same bug reported by @rauleffting. When first accessing User Management no user is loaded.
I had to follow on of the two flows so the data is loaded:
a. Badge Management -> Badge Development -> Dashboard -> User Management -> users are loaded
a. User Management -> Create New User -> users are loaded after creating new user


https://user-images.githubusercontent.com/65736872/216771738-ddeb2411-0682-4462-ae51-8a869a3ebcc5.mp4

**Comment by RafaelCastilhoCastro** [comment]
Hi @LeeSingChanYau,
I tested your PR again and observed a couple of issues I hadn't experienced before.

1. First I tryed to replicate the bug reported by @RJBornToWin. But in my case the feature worked as intended as you can see in the attached video.

https://user-images.githubusercontent.com/65736872/216772370-734b27e3-75b2-4117-9da5-abdf616c35c2.mp4


2. Now I'm experiencing the same bug reported by @rauleffting. When first accessing User Management no user is loaded.
I had to follow on of the two flows so the data is loaded:
a. Badge Management -> Badge Development -> Dashboard -> User Management -> users are loaded
a. User Management -> Create New User -> users are loaded after creating new user


https://user-images.githubusercontent.com/65736872/216772401-0c727751-aad0-4fd0-8569-dd554c9fcf8f.mp4


3. Lastly I noticed that new users with duplicate name are not being created though the app displays a confirmation/alert of its creation.


https://user-images.githubusercontent.com/65736872/216772521-ecd99b17-7710-4383-b35f-7f6c9e226f71.mp4


Thanks.

**Comment by LeeSingChanYau** [comment]
@RafaelCastilhoCastro Hello Rafael, thank you for your feedback! For issue number 2, I can take a look a look at it, but I'm not sure if my PR is causing this problem. I will try to replicate the problem. For issue number 3, yes that should be fixed. I think the problem is that the data is not refreshing correctly but I think the user was created. Now that I went to take a look at user management page I can see a user with the email foo2@bar.com.

**Comment by RafaelCastilhoCastro** [comment]
Hi @LeeSingChanYau,
You were right. I think it was my mistake about item 3. I tested again and users with duplicated names are being created normally after its related alerts. Anyway I got the same results about items 1 and 2.
Thank you.

**Comment by Filipe-Oliv** [comment]
Hi @LeeSingChanYau 

I checked your PR.

1 . For the same first and last name but different email, the yellow popup appeared...

week 03 - 02-08-23 (2)

... But when I looked for the new user it didn't show up.

week 03 - 02-08-23 (4)

2 . For the same name, last name and email the red popup appeared.

week 03 - 02-08-23 (3)

**Comment by one-community** [comment]
Thank you all for your input on this, looking forward to seeing it all integrated so we can merge it. This is one of our oldest outstanding PRs.

**Comment by angelinatr15** [comment]
Hi @LeeSingChanYau, 

I tested your PR starting from the dashboard. I proceeded to the User Management screen and saw a list of existing users. I then did the following: 
 1.  created another user with the same name and same email. 
2.  created another user with the same name and different email. 

The features looks to be working as expected. 

https://user-images.githubusercontent.com/97488347/218866606-a299de57-fb9d-4078-9a92-dc93b4fcd565.mov

**Review by angelinatr15 — DISMISSED** [review]

**Comment by vitorsaa2k** [comment]
Hello @LeeSingChanYau,

I'm on the correct branches both on Back-end and Front-end and the popup isn't showing for me, it's working just as before, as showed on the attached video:

https://user-images.githubusercontent.com/17365161/219100985-cf4001ce-f8b1-44d3-833a-c4a9b735414b.mp4

**Comment by vitorsaa2k** [comment]
Hello @LeeSingChanYau,

i made a mistake on the previous review, my apologies, now it is working as expected. If this is all the changes this PR is intended to do, i approve it.

https://user-images.githubusercontent.com/17365161/219123251-9bb707e1-7799-44c6-b6cb-a4e7abae9d0f.mp4

**Review by vitorsaa2k — DISMISSED** [review]

**Review by Filipe-Oliv — APPROVED** [review]
@LeeSingChanYau after I had seen some other people on the team running this PR I found it in one of your comments with Rafael and saw that for this PR there is a branch for BE. I tested it again and it is working as expected. My apologies.

**Comment by Theophileaseh** [comment]
Hi @LeeSingChanYau I tried testing but the user management page shows blank
Screenshot (26)

**Comment by PrabhjotSembhi** [comment]
Hey @LeeSingChanYau, This Has 3 approvals and 1 requested change. Can you please let us know? are all comments and changes are addressed

**Comment by LeeSingChanYau** [comment]
@PrabhjotSembhi, I believe they are all addressed since it works well when the right backend is being used.

**Comment by agujjar2206** [comment]
I have tested this PR and it works as expected. I created a new user with a name that has been used before and a new popup appeared asking to confirm if I want to create the user with a duplicate name. I also created a user with an email that hasn't been used before and one that has been used before. The pop-up behaved correctly.

**Review by agujjar2206 — APPROVED** [review]
Code changes look OK and follow the best standards.

**Comment by PrabhjotSembhi** [comment]
> @PrabhjotSembhi, I believe they are all addressed since it works well when the right backend is being used.

@RJBornToWin Hi, Can you please check if your comments resolved/ changes made you want?

**Comment by suigunfjc** [comment]
I tested all features by creating a new user with a duplicate name and duplicate email address. The popup works well when I'm inputting a duplicate name and everything looks good to me. I will approve it after checking the code changes. Well done!
Screenshot 2023-02-18 152800
Screenshot 2023-02-18 152833

File: src/components/UserManagement/DuplicateNamePopup.jsx
Comment (suigunfjc) [inline src/components/UserManagement/DuplicateNamePopup.jsx:38]: Hi @LeeSingChanYau, could you tell me why we are having this `DivSpacer` here? I don't see it being used in `DuplicateNamePopup`.

**Comment by Filipe-Oliv** [comment]
> Hi @LeeSingChanYau I tried testing but the user management page shows blank Screenshot (26)

Hi @Theophileaseh 

Did you run the BE? https://github.com/OneCommunityGlobal/HGNRest/pull/272

He put it in a comment section instead of the main one. Try it and see if it runs.

**Review by tingedudu8 — APPROVED** [review]
Hi @LeeSingChanYau, I tested your PR by creating a new user with a duplicate name, the correct modal shows, and the alert info regarding to the duplicate email works well. I will approve it. Thanks!
PR657
Screen Shot 2023-02-22 at 1 54 52 PM
Screen Shot 2023-02-22 at 1 54 19 PM

**Comment by harlleybastos** [comment]
Hi @LeeSingChanYau , as I said on the Back PR, I already tested and it's okay !

https://user-images.githubusercontent.com/37200778/221208553-fd6df705-3ec8-42cd-a4e8-9e3fbe9ba773.mp4

File: src/components/UserManagement/UserManagement.jsx
Comment (rauleffting) [inline src/components/UserManagement/UserManagement.jsx:238]: To delete.

**Review by rauleffting — CHANGES_REQUESTED** [review]
Hi @LeeSingChanYau. Your PR was tested and everything is working quite well, as you can see below:

https://user-images.githubusercontent.com/29555732/221361951-8a8d482c-20e3-4b70-b441-8585ebe031e2.mp4



However, there's one debug output in the code that must be deleted before we merge. May you do this?

Thank you very much!

**Comment by mpavithr** [comment]
It works for me as well - tested with backend running.

I can see the popup box when a duplicate user is created. I tested this with same duplicate name, capitalized duplicate name, mixed case duplicate name and the popup box comes up each time.
duplicate_popupbox
If I use an existing email, I also cannot proceed 
mixed_capitalization_doesn't
Once a duplicate is created, they appear in search for me as well.

**Comment by dkamghem** [comment]
It seems to work as intended. A pop out message is returned when having duplicate email preventing you from creating user. If username is duplicate a message is coming out to let you know but user is still created.
Screen Shot 2023-02-27 at 4 46 29 PM (2)
Screen Shot 2023-02-27 at 5 44 36 PM (2)
Screen Shot 2023-02-27 at 5 47 10 PM (2)

**Comment by NyahJ** [comment]
Hi @LeeSingChanYau,

The test from my side went pretty well, I used the same email address and names as suggested. Thanks!

截屏2023-02-27 下午4 33 48
截屏2023-02-27 下午4 39 58

File: src/components/UserManagement/DuplicateNamePopup.jsx
Comment (RJBornToWin) [reply src/components/UserManagement/DuplicateNamePopup.jsx:38]: Still not resolved.

File: src/components/UserManagement/UserManagement.jsx
Comment (RJBornToWin) [reply src/components/UserManagement/UserManagement.jsx:238]: Still not resolved.

File: src/components/UserProfile/AddNewUserProfile/UserProfileAdd.jsx
Comment (RJBornToWin) [inline src/components/UserProfile/AddNewUserProfile/UserProfileAdd.jsx:516]: === should be used for comparison instead of ==

Comment (RJBornToWin) [reply src/components/UserProfile/AddNewUserProfile/UserProfileAdd.jsx:516]: Can't we use find instead of filter ? Filter will run for entire length of userProfiles where as find will stop at first instance adding to performance.

File: src/components/UserProfile/AddNewUserProfile/UserProfileAdd.jsx
Comment (RJBornToWin) [inline src/components/UserProfile/AddNewUserProfile/UserProfileAdd.jsx:577]: When "if within else" gets executed then toast.success will not execute. can we change this to else if and else to improve readability.

File: src/components/UserProfile/AddNewUserProfile/UserProfileAdd.jsx
Comment (RJBornToWin) [inline src/components/UserProfile/AddNewUserProfile/UserProfileAdd.jsx:618]: Can you include break statement? When the if condition fails, the switch doesn't stop.

**Review by RJBornToWin — CHANGES_REQUESTED** [review]
Functionally correct. Able to create user with duplicate names after confirmation. 
Multiple comments added towards readability, best practices and clean code.

**Comment by RafaelCastilhoCastro** [comment]
Hi guys. Just a friendly reminder all reviewers should check the code on every modified file and approve it on Github.
@RJBornToWin thanks for the thorough review.
@LeeSingChanYau When possible please address the points brought up by RJ and will be able to move this PR to be reviewed by the leadership team.
Thank you.

**Comment by one-community** [comment]
RJ and Raul's reviews ARE the leadership team/final reviews. We're just waiting on Alan to address Raul's changes and then we can merge this.

File: src/components/UserProfile/AddNewUserProfile/UserProfileAdd.jsx
Comment (RJBornToWin) [reply src/components/UserProfile/AddNewUserProfile/UserProfileAdd.jsx:584]: Also, I find some issues in this duplicate name flow. Please go through the video to understand it.

https://www.dropbox.com/s/d3u73ls3yya0jec/Alan_PR_review.mov?dl=0

@LeeSingChanYau  Could you please test your code for all these flows before submitting for next review ? Thanks and it will reduce the time spent on repeated reviews. 

@ Everyone: These are the major testcases/usecases for this PR.
case 1. Create new user case 2. Create new user with duplicate name case a. Confirm case b.Change name 3. Filter and create user following case 1 and case 2 a, case 2 b. For all these cases, see if the user list gets updated accordingly.

Comment (LeeSingChanYau) [reply src/components/UserProfile/AddNewUserProfile/UserProfileAdd.jsx:583]: @RJBornToWin thank you for the feedback. Please review. Changed conditional statements when calling createUser. Now we return when this.checkIfDuplicate(userData.firstName, userData.lastName) && !allowsDuplicateName) is met, because a the popup will open and we dont to create the user just yet. If we allow duplicateName and
 there are no warnings then we will call toast.success().
 
 If you have a better idea on how to implement the conditions, please leave code suggestions if possible.

**Comment by RJBornToWin** [comment]
@LeeSingChanYau I have tested the application again after your recent changes. The basic flow fails. Unless all the mentioned testcases pass, this PR cannot be approved. Attaching the video here for your reference.

https://user-images.githubusercontent.com/16258136/222608239-f9e2b707-7a17-4312-9505-af3131870e4d.mov

**Comment by LeeSingChanYau** [comment]
https://user-images.githubusercontent.com/67836976/222646054-4d9e6971-4080-42c9-96bd-635ba6e19e07.mov

Hello @RJBornToWin, its working on my end. Please tell me if you still see this.

**Comment by Lunaliar** [comment]
Hey @LeeSingChanYau  I tested your PR and was able to get the proper errors for duplicate email as well as the confirmation popup for making a duplicate named user. I received the success message of the new users being added but upon searching for them I was not able to find them in the list of all users. I read through your code and it looks great! Good job!
emailError
name isalreadyUsed
successuser
User not found

**Comment by vitorsaa2k** [comment]
Not working propely.

https://user-images.githubusercontent.com/17365161/222797415-2c96111b-45fd-4df7-945b-db4727e4bb3c.mp4

1. When first creating a user with an already existing name
- I get the warning and the modal to comfirm if I want to create a user with a duplicate name.
- Choosing to change the name, and trying to create the same user again, I get the warning saying that the email is already being used. So the user was created even I choosing to not create it.
- Keeping in the user creation tab and changing the email makes the feature work as intended.

**Comment by LeeSingChanYau** [comment]
@vitorsaa2k Are you sure the user with that email was not created before you recorded the video. I can't replicate this problem.

https://user-images.githubusercontent.com/67836976/222812617-4670b7e7-0c08-4616-982c-79240ee62520.mov

**Review by rauleffting — CHANGES_REQUESTED** [review]
Hi @LeeSingChanYau. I found a bug. The form is not accepting number 6 on the inputs. Please take a look:

https://user-images.githubusercontent.com/29555732/222836413-149c600c-b2ee-4232-90dd-746675cf33c9.mp4


Also, make sure that the fields accept all characters that a user would use.

**Comment by LeeSingChanYau** [comment]
@rauleffting I think I found the problem. As you can see below all the numbers that are equal or less than length of zero will trigger this message. Even though this was not intended to be part of the PR I will fix it.
image

**Comment by LeeSingChanYau** [comment]
Please review @rauleffting.

**Comment by ayush0904** [comment]
Hi @LeeSingChanYau. I tested the functionality and it is working as intended. Thank you.

**Review by rauleffting — APPROVED** [review]
It's working. You're the best, Alan! Thank you very much!

https://user-images.githubusercontent.com/29555732/222935723-a83a5cae-8c14-4008-87a7-d38fd86777d8.mp4

**Pull request merged**
